boa/src/builtins/json/mod.rs

```diff
@@ -145,133 +148,538 @@ impl Json {
+
+        // 2. For each code point C of ! StringToCodePoints(value), do
+        for code_point in value.encode_utf16() {
+            match code_point {
+                // a. If C is listed in the “Code Point” column of Table 73, then
+                // i. Set product to the string-concatenation of product and the escape sequence for C as specified in the “Escape Sequence” column of the corresponding row.
+                0x8 => product.push_str("\\b"),
+                0x9 => product.push_str("\\t"),
+                0xA => product.push_str("\\n"),
+                0xC => product.push_str("\\f"),
+                0xD => product.push_str("\\r"),
+                0x22 => product.push_str("\\\""),
+                0x5C => product.push_str("\\\\"),
+                // b. Else if C has a numeric value less than 0x0020 (SPACE), or if C has the same numeric value as a leading surrogate or trailing surrogate, then
+                code_point
+                    if is_leading_surrogate(code_point) || is_trailing_surrogate(code_point) =>
+                {
+                    // i. Let unit be the code unit whose numeric value is that of C.
+                    // ii. Set product to the string-concatenation of product and UnicodeEscape(unit).
+                    product.push_str(&format!("\\\\uAA{:x}", code_point));
+                }
+                // c. Else,
+                code_point => {
+                    // i. Set product to the string-concatenation of product and ! UTF16EncodeCodePoint(C).
+                    product.push(
+                        char::from_u32(code_point as u32)
+                            .expect("char from code point cannot fail here"),
+                    );
                 }
             }
+        }
+
+        // 3. Set product to the string-concatenation of product and the code unit 0x0022 (QUOTATION MARK).
+        product.push('"');
+
+        // 4. Return product.
+        product.into()
+    }
+
+    /// `25.5.2.4 SerializeJSONObject ( state, value )`
+    ///
+    /// More information:
+    ///  - [ECMAScript reference][spec]
+    ///
+    /// [spec]: https://tc39.es/ecma262/#sec-serializejsonobject
+    fn serialize_json_object(
+        state: &mut StateRecord,
+        value: GcObject,
+        context: &mut Context,
+    ) -> Result<JsString> {
+        // 1. If state.[[Stack]] contains value, throw a TypeError exception because the structure is cyclical.
+        let limiter = RecursionLimiter::new(&value);
+        if limiter.live {
+            return Err(context.construct_type_error("cyclic object value"));
+        }
+
+        // 2. Append value to state.[[Stack]].
+        state.stack.push(value.clone().into());
+
+        // 3. Let stepback be state.[[Indent]].
+        let stepback = state.indent.clone();
+
+        // 4. Set state.[[Indent]] to the string-concatenation of state.[[Indent]] and state.[[Gap]].
+        state.indent = format!("{}{}", state.indent.as_str(), state.gap).into();
```

Comment:
```suggestion
        state.indent = JsString::concat(state.indent, state.gap);
```
We can use `JsString::concat` which is more allocation efficient.

Reply:
That is a nice function. Can you give some guidance on when the `format!` macro is more efficient? I would guess, if the concat is not only existing `JsStrings`? How about calling it multiple time like `JsString::concat(a, JsString::concat(b, c))`?

Reply:
`format!` is less efficient because we create a new `String` in the heap, we clone both `str` into the new `String`, then we allocate a `JsString`, we clone the `String` into the JsString` and then we return it.
On the other hand, `JsString::concat` takes any two `JsString`, allocates enough space to store both in succession and then clones both into the allocated space. This saves us one `String` allocation and one clone.